lib/auth/register.go

```diff
@@ -109,37 +109,51 @@ func Register(params RegisterParams) (*Identity, error) {
 		return nil, trace.Wrap(err)
 	}
 
-	// Attempt to register through the auth server, if it fails, try and
-	// register through the proxy server.
-	ident, err := registerThroughAuth(token, params)
-	if err != nil {
-		// If no params client was set this is a proxy and fail right away.
-		if params.GetHostCredentials == nil {
-			log.Debugf("Missing client, failing with error from Auth Server: %v.", err)
-			return nil, trace.Wrap(err)
-		}
-
-		log.Errorf("Failed to register through auth server: %v; falling back to trying the proxy server", err)
+	type registerMethod struct {
+		call func(token string, params RegisterParams) (*Identity, error)
+		desc string
+	}
+	registerThroughAuth := registerMethod{registerThroughAuth, "with auth server"}
+	registerThroughProxy := registerMethod{registerThroughProxy, "via proxy server"}
+
+	var registerMethods []registerMethod
+	if params.GetHostCredentials == nil {
+		log.Debugf("Missing client, it is not possible to register through proxy.")
```

Comment:
This sounds like a warning to me unless this is customary to have no client set, I'd hint at this if this is in fact noteworthy (`log.Warn`).

Reply:
FWIW, this was a debug level message in the original code as well. The missing `GetHostCredentials` just causes the proxy registration method to be not considered as viable while the other pathway remains untouched by this. Therefore it doesn't seem reasonable to clutter the log with warnings saying "not possible to register through proxy" when there needed to be no intent of going through proxy at all in the first place.